nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java

```diff
@@ -0,0 +1,257 @@
+import static com.uber.nullaway.NullabilityUtil.getAnnotationValueArray;
+
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.LiteralTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.code.Symbol;
+import com.uber.nullaway.ErrorMessage;
+import com.uber.nullaway.NullAway;
+import com.uber.nullaway.NullabilityUtil;
+import com.uber.nullaway.Nullness;
+import com.uber.nullaway.dataflow.AccessPath;
+import com.uber.nullaway.dataflow.AccessPathNullnessPropagation;
+import com.uber.nullaway.dataflow.NullnessStore;
+import com.uber.nullaway.handlers.AbstractFieldContractHandler;
+import com.uber.nullaway.handlers.MethodAnalysisContext;
+import com.uber.nullaway.handlers.contract.ContractUtils;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.VariableElement;
+import org.checkerframework.nullaway.dataflow.cfg.node.MethodInvocationNode;
+
+/**
+ * This Handler parses {@code @EnsuresNonNullIf} annotation and when the annotated method is
+ * invoked, it marks the annotated fields as not-null in the following flows.
+ */
+public class EnsuresNonNullIfHandler extends AbstractFieldContractHandler {
+
+  // Set to true when the current visited EnsuresNonNullIf method has correct semantics
+  private boolean semanticsHold;
+  // List of fields missing in the current EnsuresNonNullIf method so we can build proper error
+  // message
+  @Nullable private Set<String> missingFieldNames;
+
+  public EnsuresNonNullIfHandler() {
+    super("EnsuresNonNullIf");
+  }
+
+  /**
+   * Validates whether all parameters mentioned in the @EnsuresNonNullIf annotation are guaranteed
+   * to be {@code @NonNull} at exit point of this method.
+   */
+  @Override
+  protected boolean validateAnnotationSemantics(
+      MethodTree tree, MethodAnalysisContext methodAnalysisContext) {
+    // If no body in the method, return false right away
+    // TODO: Do we need proper error message here?
+    if (tree.getBody() == null) {
+      return false;
+    }
+
+    // clean up state variables, as we are visiting a new annotated method
+    semanticsHold = false;
+    missingFieldNames = null;
+
+    // We force the nullness analysis of the method
+    NullAway analysis = methodAnalysisContext.analysis();
+    VisitorState state = methodAnalysisContext.state();
+    analysis
+        .getNullnessAnalysis(state)
+        .forceRunOnMethod(new TreePath(state.getPath(), tree), state.context);
+
+    // If listed fields aren't checked for their nullness, return error
+    if (!semanticsHold) {
```

Comment:
I think reporting errors here again won't quite work in the presence of nested methods.  Instead, I suggest keeping another map from method trees to the information from the annotation (the field names and whether they should be non-null when returning `true` or `false`).  Then, we should separately check each `return` statement in `onDataflowVisitReturn` and report an error directly at that point.  (This will also give better error locations.)  Right now the `VisitorState` (needed to call `reportMatch`) is not available as a parameter in `onDataflowVisitReturn`, but it should be easy to pass in if you add it as an additional parameter (it's available in a field [at this call site](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java#L855-L855)).

Reply:
Oh I only saw this comment now. Will try to move the validation to each call to the onDataflowVisitReturn. 

Not sure I will need the map, as I believe I already have everything I need there, besides the VisitorState, which I'll try to pass it on.